client/http-server/src/user_data.rs

```diff
@@ -0,0 +1,208 @@
+use codec::Decode;
+use frame_support::storage::generator::StorageDoubleMap;
+use hyper::Uri;
+use pallet_lfs_user_data as pallet;
+use sc_client::Client;
+use sc_client_api::{backend, CallExecutor};
+use sp_core::crypto::Ss58Codec;
+use sp_core::storage::StorageKey;
+use sp_lfs_core::{LfsId, LfsReference};
+use sp_runtime::{generic::BlockId, traits::Block as BlockT};
+use std::marker::PhantomData;
+use std::sync::Arc;
+
+use crate::traits::Resolver;
+
+#[derive(Clone)]
+enum NextResolveStep {
+	UserData,
+	Glob,
+	NotFound,
+	End,
+}
+
+impl NextResolveStep {
+	fn next(&self) -> Self {
+		match self {
+			NextResolveStep::UserData => NextResolveStep::Glob,
+			NextResolveStep::Glob => NextResolveStep::NotFound,
+			NextResolveStep::NotFound => NextResolveStep::End,
+			_ => NextResolveStep::End,
+		}
+	}
+}
+
+pub struct UserDataResolveIterator<L, B, E, Block: BlockT, RA, T: pallet::Trait> {
+	client: Arc<Client<B, E, Block, RA>>,
+	best_block: BlockId<Block>,
+	root_key: T::AccountId,
+	_marker: PhantomData<(T, L)>,
+	uri: Uri,
+	next: NextResolveStep,
+}
+
+impl<L, B, E, Block, RA, T> UserDataResolveIterator<L, B, E, Block, RA, T>
+where
+	B: backend::Backend<Block>,
+	E: CallExecutor<Block>,
+	Block: BlockT,
+	L: LfsId,
+	T: pallet::Trait,
+{
+	fn new(
+		client: Arc<Client<B, E, Block, RA>>,
+		best_block: BlockId<Block>,
+		root_key: T::AccountId,
+		uri: Uri,
+	) -> Self {
+		Self {
+			client,
+			best_block,
+			root_key,
+			uri,
+			next: NextResolveStep::UserData,
+			_marker: Default::default(),
+		}
+	}
+
+	fn lookup(&self, key: &StorageKey) -> Option<L> {
+		self.client
+			.storage(&self.best_block, key)
+			.map(|o| {
```

Comment:
You can use `and_then` instead of `.map().flatten()`. Even better to just use `?` most likely :)

Reply:
:+1: on the and_then, but not for the `?`:? The lookup and how it is being used not finding it is not considered an error but a common expected case, thus keeping it as an `Option`.

Reply:
`?` actually works for `Option`! :)